### DISCUSSION TITLE
webui working with CuBLAS / Offloading GPU Layers

### DISCUSSION THREAD

**Discussion by Daeformator** [oobabooga/text-generation-webui#2339] [Q&A] [answered]
How do I get this going to work, with llamacpp I normally can see:
llama_model_load_internal: [cublas] offloading 35 layers to GPU
llama_model_load_internal: [cublas] total VRAM used: 5956 MB

or something similar during the load up, when I'm going through oobabooga, it doesn't do this even when I put  --n-gpu-layers 35 in the webui CMD_RUN section... 

Anything I'm missing?

**Comment by BetaDoggo** [comment, marked as answer]
You need to compile llama-cpp-python with cublas support as explained on the [wiki](https://github.com/oobabooga/text-generation-webui/blob/main/docs/llama.cpp-models.md). As far as I know this is broken right now.

**Reply by Daeformator** [reply]
Thanks for the reply :)

**Reply by viperwasp** [reply]
Can anyone confirm if this is working right now? Or still broken? It won't work for me even doing the windows setup steps.

**Reply by ossirytk** [reply]
I just run the current develop and it works. There might be an issue in the llamma-cpp-python instructions where the set command for environment variables was not the correct if you are running powershell. I think someone mentioned it in the issues somewhere. I run cuda powershell so I tried this.
`$env:FORCE_CMAKE=1
$env:CMAKE_ARGS="-DLLAMA_CUBLAS=on"
pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir
`

**Reply by viperwasp** [reply]
Thanks what version of Visual Studios are you using?

**Reply by ossirytk** [reply]
2022 community edition. I just installed it since it was a Cuda toolkit dependency. I previously had installed VisualStudio BuildTools, but I had to delete it since it conflicted with the Cuda installation. I needed to install the Cuda Toolkit with extended options to get the VisualStudio cmake bindings to work. Then I had to delete old build caches since they pointed to the old BuildTools cmake. It was kind of a hassle, but the perf boost is pretty significant for me.

**Reply by viperwasp** [reply]
The reason I'm asking is I am following these steps. 
Ok, now I was able to install GPU mode with these steps:

    I uninstalled all Visual Studio and installed [VS 2019 BuildTools](https://my.visualstudio.com/Downloads?q=visual%20studio%202019&wt.mc_id=o~msft~vscom~older-downloads)
    add to PATH this line: C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64 
    I had to install the latest [Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/) separately (not enough VS version!), including the ones called Desktop C++
    I had to copy four files from this folder: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\extras\visual_studio_integration\MSBuildExtensions to this folder: C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Microsoft\VC\v160\BuildCustomizations
    then I ran the four commands above.
    I'm going to try it :)
    
    From this link 
    https://github.com/oobabooga/text-generation-webui/discussions/2488

**Reply by viperwasp** [reply]
Okay I just read your reply I may need to follow your steps of what you just said and get 2022 community and do all of that. lol
I just had 2022 community.  I don't use powershell I have it installed but I don't remember opening it before? Do I need to do anything with it or just use CMD?

**Reply by ossirytk** [reply]
It seems that on Windows the Cuda toolkit can't update the binding correctly if you have both the BuildTools and visual studio installed. This might also occur if you have multiple versions of VisualStudio. But VSCode makes the BuildTools kind of redundant, so you can just uninstall those. I don't think the default installation includes the updates to VSCode extensions either. You have to select modified installation and enable those.

**Reply by viperwasp** [reply]
I'm really sorry but can you explain all this in English for my smaller brain? lol 
I know what Cuda toolkit is.... Which I have a question I had 12.1 installed I downgraded to 11.7 due to an Ooba red error stating a compatibility issue. So which Cuda Tookit should I have installed.

Basically if you use windows and even more so windows 10. I want to copy your setup to see if it works? 
So I should uninstall all of Visual Studios. Everything to do with it? And then what? Install 2022 community? What custom options during the Visual Studio setup should I check off? 

Second windows SDK I think it's called? What version of that do I need? 
I did not know that Buildstools was separate of Visual Studios I thought you put in Visual Studios to get the builds tools? So I am very lost? What are the Buildstools? What is that called? How do I get it? 

Finally I tried everything I've read suggested. I even used CMD to try three different BLAS backends. I even copied and pasted over those four files that the link I provided you suggesed from that one person. Files from the Nvidia Cuda to the Visual Studios. Nothing works. In my CMD when I run OOBA GGML it says BLAS=0.... I have a question does BLAS=0 mean that it's not working at all? Or is there still a setting in OOBA I need to do. I was able to get Quant_Cuda installed. I don't get any errors anymore at all. It's just it won't load GGML into Video Memory at all. 0% of it. Regardless were I put the GPU layers etc. I change both of those sliders that are to do with layers etc. BLAS has never been anything except 0. I feel like I'm missing something simple.  I know about setting up paths I don't think I'm having that issue. Unless I have to set the path in the "start_windows.bat" if you think I need to change anything in "start_windows" let me know? 

If you can't help that is fine. 

RTX 4080 (Newest Drivers) 
12700K 
Windows 10 Home. 
Ooba is fully updated.

**Comment by ossirytk** [comment]
I think people may be confusing the pre-layer thing with the llama.cpp cuda support. These are two separate things. The pre-layers is a booga thing and I'm under the impression that it's slower. It also requires pytorch or whatever. You don't really need this if you run the bin files. The llama.cpp should only require that you install Coda toolkit and build with the CUBLAS option. Then you set the n_layers or whatever it was to some number. This should require no other settings and works with the current bin models. 

I use the latest community Visual studio and the latest Cuda toolkit. You should make sure that only one 64x version of VisualStudio exist when you run the Toolkit. You should make sure that no buildtools exist before you run the toolkit. You need to run the "custom" installation and the make sure that the VisualStudio cmake is in path. You might need to clean previous build files. Then update the ooba build with the CUBLAS enabled. 

Open Task manager and check you core number. Not the thread number, but the core number. Go to the gpu page and keep it open. Run the server and go to the model tab. Load a 13b quantized bin type GGMLmodel. Set thread count to match your core count. 1thread/core is supposedly optimal. Set n-gpu-layers to 20. Run the chat. You should see gpu being used. If gpu is 0 then the CUBLAS isn't enabled.

**Reply by viperwasp** [reply]
Thanks that's basically what I'm doing as far as the settings are on concerned. Believe me I tried every setting I can think of. Blas always = 0 and GPU memory never increases. But were you helped me the most so far is the dependances. I did most or all of what you said but it's possible I needed a clean build etc. Uninstall Visual studios and Cuda toolkit etc. Then maybe put them in and then install ooba. 

I have a question if you can answer. Are you using windows 10 with this enabled? Either way what version of Visual Studios do you recommend? 2022 community? What version of Cuda Toolkit? Most recent? Also does the order I install either of the matter? Thanks.

**Comment by ossirytk** [comment]
I'm running on Windows 11 with latest Visual Studio community version and the latest cuda toolkit. The only extra installation I have done is the pythorch and I'm pretty sure that I don't need it. I think most of the extra dependencies are not really needed if you only intend to run a chatbot and the can easily result in conflicts. As for the pre-layers, bitwise etc. I'm not really sure that these are up to date and actually give any significant benefit to performance. You can get the GMML QxK bin files from HuggingFace anyway. I would not use the ooba for training or anything like that. I do that stuff in terminal.

**Reply by viperwasp** [reply]
Resolved see https://github.com/oobabooga/text-generation-webui/discussions/2635#discussioncomment-6162790
